ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/commands/tracker/CommandTracker.scala

```diff
@@ -279,33 +292,70 @@ private[commands] class CommandTracker[Context](
       }
 
       private def getOutputForCompletion(completion: Completion) = {
-        val (commandId, errorText) = {
+        val (potentialCommandKey, errorText) = {
           completion.status match {
             case Some(StatusProto(code, _, _, _)) if code == Status.Code.OK.value =>
-              completion.commandId -> "successful completion of command"
+              TrackedCommandKey(
+                completion.submissionId,
+                completion.commandId,
+              ) -> "successful completion of command"
             case _ =>
-              completion.commandId -> "failed completion of command"
+              TrackedCommandKey(
+                completion.submissionId,
+                completion.commandId,
+              ) -> "failed completion of command"
+          }
+        }
+
+        val commandKey = if (potentialCommandKey.submissionId.isEmpty) {
+          val potentialKeys =
+            pendingCommands.keys.filter(_.commandId == potentialCommandKey.commandId)
+          if (potentialKeys.size > 1) {
+            throw new IllegalStateException(
+              s"There are multiple pending commands for the id ${potentialCommandKey.commandId}. This can only happen for the mutating schema."
+            ) with NoStackTrace
```

Comment:
A bit worried about the throw though. From what I see it's part of the graph stage, so this will be propagated to the stream, so it might be that we will lose internal state depending on the error handling strategy set on the stream.

Reply:
I can return this as a completion failure with an internal error. WDYT?

Reply:
Sounds good to me.

The problem is how do we complete all of them and remove them from pending? Do we complete just one? Which one in that case? 
Do we remove all of them? What happens when we get a completion that has no pending command in that case

Reply:
The test failures seem to be related to this actually. 
In the old code we just ignore completions that didn't have a pending command. And that makes sense as not all completions are tracked.
In the new code we assume that there is always a tracked command for a completion that doesn't have a submission id. 

In this case, when not having a submission id we should get all the pending commands for that command id and complete them with an INTERNAL failure, and remove them from the pending commands.

Reply:
> I can return this as a completion failure with an `INTERNAL` error. WDYT?

I like that idea better than an exception. `INTERNAL` works for now, but we may want to revisit this later (e.g., change it to `ABORTED` as part of the error code cleanup).


> Which one in that case?

If the application sets the same submission ID for several submissions, then it's the application's fault if it can't distinguish between them. We could also silently drop the request that reuses the submission ID, but that's a worse UX than signalling that something is off.



> What happens when we get a completion that has no pending command in that case

This can happen during regular operations if there are parallel submissions, some of which go through the `CommandService` and some hit the `CommandSubmissionService` directly. Again, if submission IDs are not unique, then the application will observe strange behaviour.

Reply:
Thanks, done

Reply:
Thanks, done